### PR TITLE
add data adapter for chirpstack

### PR DESCRIPTION
**Pull request by DeadByElpy**

### CODE REVIEW

File: src/adapters/index.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+};
+
+// fastify force all headers in lowercase
+const headerName = 'ns-adapter';
+
+
+export default function adaptData ( request, reply ) {
+    const adapterName = request.headers[headerName];
+
+    if ( !adapterName && adapters[adapterName] ) {
```
Comment (Pnom) [inline src/adapters/index.js]: wrong condition. It's ok if no adapter name is provided.

`if ( !adapterName ) {`
`    return true;`
`}`
`const adapter = adapters[adapterName];`
`if ( !adapter ) {`
`     error!!!!!!!!!!!!!!`
`}`

File: src/adapters/index.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        reply.sendError(errors.BAD_REQUEST, `Unsupported adapter "${adapterName}"`);
+
+        return false;
+    }
+
+    const adapter = adapters[adapterName];
+    const body = adapter(request, this.log);
+
+    if ( !body ) {
```
Comment (Pnom) [inline src/adapters/index.js]: You don't use the adapter function result

File: src/adapters/chirpstack.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    log.trace('received from chirpstack: %j', body);
+
+    if ( !direction ) {
+        // ignore other events except uplink
+        return null;
+    }
+
+    const {deviceInfo} = body;
+    const protocol = deviceInfo?.tags?.protocol;
+    const deviceEUI = deviceInfo.devEui;
```
Comment (Pnom) [inline src/adapters/chirpstack.js:25]: may be we need `?` 
like 
`const deviceEUI = deviceInfo?.devEui;`

File: src/adapters/index.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+const adapters = {
+    chirpstack
+};
+
+// fastify force all headers in lowercase
+const headerName = 'ns-adapter';
+
+
+export default function adaptData ( request, reply ) {
```
Comment (Pnom) [inline src/adapters/index.js]: is it the correct form for the Fastify hook?
as I can see from docs there are variants
`fastify.addHook('preValidation', (request, reply, done) => {`
`fastify.addHook('preValidation', async (request, reply) => {`

File: src/adapters/chirpstack.js (hunk 3 above)
Comment (DeadByElpy) [reply src/adapters/chirpstack.js:25]: it is always present, and if not, it's a wrong device

File: src/adapters/index.js (hunk 1 above)
Comment (DeadByElpy) [reply src/adapters/index.js]: fixed

File: src/adapters/index.js (hunk 2 above)
Comment (DeadByElpy) [reply src/adapters/index.js]: fixed, also added a test suite for chirpstack adapter as separate commit

File: src/adapters/index.js (hunk 4 above)
Comment (DeadByElpy) [reply src/adapters/index.js]: fixed

**Review by Pnom — APPROVED** [review]

**Pull request merged**